### PR TITLE
Drag & drop Protractor test - PoC

### PR DESCRIPTION
**Pull request by zachlysobey**
I put this together before seeing @jmcollin78's comment on https://github.com/angular-ui-tree/angular-ui-tree/pull/538.

In short, this code is gross, and should not be merged, but it does show a successful drag+drop e2e test with protractor. 

When I get a chance, I'll keep iterating on this, but any help or advice is much appreciated. Its been a few months since I was writing WebDriver stuff so I feel quite rusty. I'm also quite new to this OSS stuff, so please give me criticism if my etiquette is off or there is anything I can do to be better at this.

### CODE REVIEW

File: e2e/basic-example/basic-example.js [hunk 1]
```diff
@@ -33,7 +33,24 @@ describe('the Basic example page', function () {
 
   describe('the dragging and dropping of nodes', function () {
     it('should allow moving a node below another node', function () {
-      // TODO
+      var handles = basicExamplePage
```
Comment (Voles) [inline e2e/basic-example/basic-example.js]: I suggest adding a method `firstTreeItemHandler()` method to the page-object (PO) since you're re-using after moving the node. The logic to find the first node is irrelevant for the test, so moving it to the PO makes the test less verbose and thus easier to read.

File: e2e/basic-example/basic-example.js [hunk 2]
```diff
@@ -33,7 +33,24 @@ describe('the Basic example page', function () {
...
+        .rootNodes.get(1)
+        .all(by.repeater('node in node.nodes'))
+        .all(by.css('[ui-tree-handle]'));
+
+      handles.get(0).getText().then(function(nodeTextBeforeDrag) {
```
Comment (Voles) [inline e2e/basic-example/basic-example.js]: Suggest renaming to `originalLabelOfFirstTreeItem`.

File: e2e/basic-example/basic-example.js [hunk 3]
```diff
@@ -33,7 +33,24 @@ describe('the Basic example page', function () {
...
+
+        browser.actions().dragAndDrop(
+          handles.get(0),
+          { x: 0, y: 200 }
+        ).perform();
+
+        handles.get(0).getText().then(function(nodeTextAfterDrag) {
```
Comment (Voles) [inline e2e/basic-example/basic-example.js]: Suggest renaming to `currentLabelOfFirstTreeItem`.

File: e2e/basic-example/basic-example.js [hunk 4]
```diff
@@ -33,7 +33,24 @@ describe('the Basic example page', function () {
...
+          console.log(nodeTextBeforeDrag, nodeTextAfterDrag);
+          expect(nodeTextBeforeDrag).not.toBe(nodeTextAfterDrag);
```
Comment (Voles) [inline e2e/basic-example/basic-example.js]: Instead of checking if the label does not equal the original label, I would explicitly check the values of both labels here. That would result in 2 expects:

**Before the move**
```
expect(firstNodeLabel).toBe('node1');`
expect(secondNodeLabel).toBe('node2');`
```

**After moving the node**
```
expect(firstNodeLabel).toBe('node2');
expect(secondNodeLabel).toBe('node1');
```

**Comment by Voles** [comment]
First of all, thanks very much. I really appreciate your work!

I'm no expert with E2E-tests, but I've added my feedback. In general, I think it's good to move as much element selectors as possible to the page-object file.

Looking forward to see how this progresses :smiley:

**Comment by zachlysobey** [comment]
Thanks for the thorough feedback! I definitely agree that a more full-featured PO would be better. I was actually considering a more full featured API for accessing the different nodes.

Possible usage: `basicExamplePage.node('node1.1').handle()`

Hopefully I can dig in a bit more this weekend.

**Comment by zachlysobey** [comment]
Other crazy ideas, but maybe too difficult to implement:

 - `basicExamplePage.node(by.text('node1.1')).handle()`
 - `basicExamplePage.node(by.position(0, 1)).handle() // first rootNode, 2nd child`

**Comment by Voles** [comment]
Any updates on this @zachlysobey? Would be nice if we could merge your initial setup for testing D&D! :tada:  Let me know if you need further help with this.

**Comment by zachlysobey** [comment]
Unfortunately, I've been pretty sick all week and haven't been able to really dig in. I have a couple projects that need precedence but I'm sure I can have something more solid by next weekend.

**Comment by zachlysobey** [comment]
Ok @Voles, I think I finally had a bit of a breakthrough.

Still a lot of cleanup (and writing some tests) to do, but I came up with what I think is a reasonable implementation for `basicExamplePage.getNodeAtPosition` by using xpath.

I struggled quite a while trying to get it to work with nested `by.repeater` stuff, and then with `by.css` but I was either being dense or its very difficult or impossible to reliably look up these elements using those methods.

In any case, just wanted to post some progress here for you to see if you're curious. I expect I'll have something worth merging in by the end of the coming weekend.

**Comment by zachlysobey** [comment]
@Voles I'll finish up those last couple tests soon. Anything else that you want done before we can merge this in? I have a bunch of improvements I'd *like* to make, but will feel good to get this merged in as soon as it feels good enough and provides some value.

**Comment by Voles** [comment]
@zachlysobey this looks really great and useful! I love the `.getNodeAtPosition(2, 1)`, it's intuitive and the tests are actually very readable.

I would like to finish up the refactoring to remove the scope dependency (#648) soon. Your work will be helpful to test if the behaviour is still te same. And can easily add new tests :+1:

**Comment by zachlysobey** [comment]
I just spent a long time trying (and failing) to get something working for those last 2 "TODO" tests:
 - 'should allow adding a node to another node to make it a child-node'
 - 'should allow removing a child-node from a node to put it on the same tree level'

I also tried without luck to do something better with the magic numbers I use to do the drag+drop. (`{ x: 0, y: 200 }`), but could not quite figure out how to do what I wanted.

In short, I think I've exhausted my patience on this for a while.

 Maybe at some point in the future we can work together on this in a live-chat sort-of situation and figure this out? A bit more insight into how this thing works under the hood might help, and of course having someone to bounce ideas off in real time is always good.

**Comment by Voles** [comment]
Looks great, @zachlysobey! :+1: 

I'll merge the current state of the art of your PR. Can you create a separate PR with what you already have for those two tests with a TODO? I can review them separately and think with you on how to write those tests.

If we can't make it work that way, we can schedule a Hangouts or Skype session.

**Comment by zachlysobey** [comment]
@Voles: awesome! Really happy to see this merged.

Regarding the progress on the other TODO tests, I ended up actually throwing out everything I did. It was a series of many 10-20 minute failed experiments, none of which looked like they'd provide any actual benefit.

I definitely want to dig in again, and the hangout/skype session might be very helpful -- but lets put that off till some point in the new year. Perhaps just a new *issue* now for the missing tests?
